pdp/ast/jast/obligations.go

```diff
@@ -24,11 +25,43 @@ func (ctx context) unmarshalObligationItem(d *json.Decoder) (pdp.AttributeAssign
 			return newUnknownAttributeError(k)
 		}
 
-		e, err = ctx.unmarshalValueByType(a.GetType(), d)
+		delim, val, err := jparser.CheckObjectStartOrValue(d, "argument")
 		if err != nil {
-			return bindError(err, k)
+			return err
+		}
+		if delim == "" {
+			e, err = ctx.unmarshalValueByTypeObject(a.GetType(), val)
+			if err != nil {
+				return err
+			}
+		} else if delim == jparser.DelimObjectStart {
+			e, err = ctx.unmarshalExpression(d)
+			if err != nil {
+				return bindError(err, k)
+			}
+		} else if delim == jparser.DelimArrayStart {
+			val, err := jparser.GetArray(d, "obligations")
```

Comment:
JAST is a stream parser. You need to work on token level. Don't try to get whole array and then parse it. Instead parse items as they appear with `CheckArrayStart` and by token iterations. However, if we follow no backward compatibility this isn't needed. Because obligation is always expression and you can just use existing expression parser here.

Reply:
Yes, totally understood. (please also see https://github.com/infobloxopen/themis/pull/112#discussion_r164430421)